### DISCUSSION TITLE
Mongoose: Cannot overwrite "model" &&  mongoose.models is not a function errors

### DISCUSSION THREAD

**Discussion by Klkrieg** [vercel/next.js#17252] [Help]
I am new to next and having a hard time with mongoose in my next js project. 

I have my api pages set up and working correctly. I am using a similar set up as in [with-mongodb-mongoose](https://github.com/vercel/next.js/tree/canary/examples/with-mongodb-mongoose).

here is my dbConnect
```
import mongoose from "mongoose";

const connection = {};
const dbConnect = async () => {
	if (connection.isConnected) {
		return;
	}
	try {
		const db = await mongoose.connect(process.env.MONGODB_URI, {
			useNewUrlParser: true,
			useUnifiedTopology: true,
		});
		connection.isConnected = db.connections[0].readyState;

		console.log(connection.models.Resource);
	} catch {
		return "Did not connect to Mongo";
	}
};

export default dbConnect;
```
here is my schema
```
const resourceSchema = new mongoose.Schema({
	key: value
});

const Resource = mongoose.model("Resource", resourceSchema);

export default Resource;
```

When writing a normal /api/resource endpoint, the get request works just fine.
```
import Resource from "../../../models/resource";
import dbConnect from "../../../utils/dbConnect";

dbConnect();

export default async (req, res) => {
	const { method } = req;
	switch (method) {
		case "GET":
			try {
				const resource = await Resource.find({});
				res.status(200).json({ success: true, data: resource });
			} catch {
				res.status(400).json({ success: false });
			}
	}
};
```
However when trying to `getStaticProps()` to make the request
```


export async function getStaticProps() {
	dbConnect();
	const data = await Resource.find({}).then(res => res.json())

	return {
		props: {
			data,
		},
	};
}

```
 it cannot read the mongoose code. This is resulting in the `Cannot overwrite "Data" model`.

 I have tried fixing it with
 `const data = mongoose.models.data || mongoose.model("Data", dataSchema)`
 in my data.js model. 
However, this then fires the error  
`mongoose.models is not a function`.

I am not fetching the endpoint in `getStaticProps()`. I am writing this request directly in this page export. 

I cannot find anything anywhere on this error. All help is greatly appreciated.

**Comment by Danielioanitescu92** [comment]
Have you solved it? I'm stuck with the same issue. I tried cloning other github reps with next js and mongoose boilerplates, different styles, still same error.
Even posted my question on stackoverflow, no answers yet.
Let me now if you found the issue, please.

**Comment by lafisrap** [comment]
Here I found help:
https://stackoverflow.com/questions/60660488/is-it-possible-to-run-mongoose-inside-next-js-api/65371870

**Comment by drscottlobo** [comment]
I found this solution which currently works for my project using mongoose: 

`import mongoose from 'mongoose';
const Schema = mongoose.Schema;

const userSchema = new Schema({
    username: {type: String, unique: true},
    name: {
        first: {
            type: String, 
            trim: true,
            required: true
        },
        last: {
            type: String, 
            trim: true,
            required: true
        }
    },
});

mongoose.models = {};

export default mongoose.model('User', userSchema);`

**Reply by mrylraihan** [reply]
This works great but all i had to do is run my build again and it worked 

const Task = mongoose.models.Task || mongoose.model('Task', taskSchema)
export default Task
